File: Rust/2021/25.rs
#![feature(test)]

#[derive(Copy, Clone, Eq, PartialEq)]
enum Type {
    Empty,
    Right,
    Down,
}

type Input = Vec<Vec<Type>>;

fn setup(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            line.chars()
                .map(|c| match c {
                    'v' => Type::Down,
                    '>' => Type::Right,
                    _ => Type::Empty,
                })
                .collect()
        })
        .collect()
}

fn part1(input: &Input) -> String {
    let mut grid = input.clone();
    let mut cnt = 0;
    loop {
        let mut moved = false;
        let mut new_grid = grid.clone();
        for (i, line) in grid.iter().enumerate() {
            for (j, &c) in line.iter().enumerate() {
                if c == Type::Right {
                    let k = (j + 1) % line.len();
                    if grid[i][k] == Type::Empty {
                        new_grid[i][k] = Type::Right;
                        new_grid[i][j] = Type::Empty;
                        moved = true;
                    }
                }
            }
        }
        grid = new_grid;
        let mut new_grid = grid.clone();
        for (i, line) in grid.iter().enumerate() {
            for (j, &c) in line.iter().enumerate() {
                if c == Type::Down {
                    let k = (i + 1) % grid.len();
                    if grid[k][j] == Type::Empty {
                        new_grid[k][j] = Type::Down;
                        new_grid[i][j] = Type::Empty;
                        moved = true;
                    }
                }
            }
        }
        grid = new_grid;
        cnt += 1;
        if !moved {
            break;
        }
    }
    cnt.to_string()
}

aoc::main!(2021, 25, ex: 1);

File: Python/2021/25.py
from lib import *

input = read_input(2021, 25)

lines = input.splitlines()


right = set()
down = set()
width = len(lines[0])
height = len(lines)
for i, line in enumerate(lines):
    for j, c in enumerate(line):
        match c:
            case "v":
                down.add((j, i))
            case ">":
                right.add((j, i))

i = 0
while True:
    old = {*right}, {*down}
    right = {
        new_pos if (new_pos := ((x + 1) % width, y)) not in right and new_pos not in down else (x, y) for x, y in right
    }
    down = {
        new_pos if (new_pos := (x, (y + 1) % height)) not in right and new_pos not in down else (x, y) for x, y in down
    }
    i += 1
    if (right, down) == old:
        break

print(i)
